File: Bronze/2016/US Open/03_field_reduction/FieldReduction.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.StringTokenizer;

/**
 * @author Red
 * @version 1.0
 */
public class FieldReduction {
    static int n;
    static List<List<Integer>> cowPositions = new ArrayList<>();
    static HashSet<List<Integer>> candidates = new HashSet<>();

    public static void main(String[] args) throws Exception {
        BufferedReader r = new BufferedReader(new FileReader("reduce.in"));
        PrintWriter pw = new PrintWriter("reduce.out");
        n = Integer.parseInt(r.readLine());
        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(r.readLine());
            List<Integer> cow = new ArrayList<>();
            cow.add(Integer.parseInt(st.nextToken()));
            cow.add(Integer.parseInt(st.nextToken()));
            cowPositions.add(cow);
        }
        // 四个方向找出至多 16 个候选点 使用 HashSet 去重
        cowPositions.sort((a, b) -> a.get(0) - b.get(0));
        for (int i = 0; i < 4; i++) {
            candidates.add(cowPositions.get(i));
            candidates.add(cowPositions.get(cowPositions.size() - i - 1));
        }
        cowPositions.sort((a, b) -> a.get(1) - b.get(1));
        for (int i = 0; i < 4; i++) {
            candidates.add(cowPositions.get(i));
            candidates.add(cowPositions.get(cowPositions.size() - i - 1));
        }
        // 将 HashSet 转换为 List 便于遍历
        List<List<Integer>> candidatesList = new ArrayList<>(candidates);
        // 从候选点中去除三个点计算最小面积 虽然三重循环但是candidateList.size()最大为16 因此复杂度为 O(1)
        long ansArea = 40000 * 40000;
        for (int i = 0; i < candidatesList.size(); i++) {
            for (int j = i + 1; j < candidatesList.size(); j++) {
                for (int k = j + 1; k < candidatesList.size(); k++) {
                    long[] minAndMax = findMinAndMaxWithAbsent(candidatesList, i, j, k);
                    long area = (minAndMax[1] - minAndMax[0]) * (minAndMax[3] - minAndMax[2]);
                    ansArea = Math.min(ansArea, area);
                }
            }
        }
        pw.println(ansArea);
        pw.close();
    }

    /**
     * 找出除了三个点之外的最小和最大值
     */
    public static long[] findMinAndMaxWithAbsent(List<List<Integer>> a, int index1, int index2, int index3) {
        int minX = Integer.MAX_VALUE;
        int maxX = 0;
        int minY = Integer.MAX_VALUE;
        int maxY = 0;
        for (int i = 0; i < a.size(); i++) {
            if (i == index1 || i == index2 || i == index3) {
                continue;
            }
            minX = Math.min(minX, a.get(i).get(0));
            maxX = Math.max(maxX, a.get(i).get(0));
            minY = Math.min(minY, a.get(i).get(1));
            maxY = Math.max(maxY, a.get(i).get(1));
        }
        return new long[]{minX, maxX, minY, maxY};
    }
}

File: Bronze/2016/US Open/03_field_reduction/field_reduction.py
fin = open('reduce.in', 'r')
fout = open("reduce.out", "w")
n = int(fin.readline())
x = []
y = []
min_x, second_min_x, second_min_y, min_y = 40000, 40000, 40000, 40000
second_max_x, max_x, second_max_y, max_y = 0, 0, 0, 0
# 更新最小、次小、最大、次大的x和y坐标
def update_min_max(x, y):
    global min_x, second_min_x, second_max_x, max_x, min_y, second_min_y, second_max_y, max_y
    if x < min_x:
        second_min_x = min_x
        min_x = x
    elif x < second_min_x:
        second_min_x = x

    if x > max_x:
        second_max_x = max_x
        max_x = x
    elif x > second_max_x:
        second_max_x = x

    if y < min_y:
        second_min_y = min_y
        min_y = y
    elif y < second_min_y:
        second_min_y = y

    if y > max_y:
        second_max_y = max_y
        max_y = y
    elif y > second_max_y:
        second_max_y = y

for _ in range(n):
    xi, yi = map(int, fin.readline().split())
    x.append(xi)
    y.append(yi)
    update_min_max(xi, yi)

ans = (max_x - min_x) * (max_y - min_y)

for i in range(n):
    cur_min_x = min_x if x[i] != min_x else second_min_x
    cur_max_x = max_x if x[i] != max_x else second_max_x
    cur_min_y = min_y if y[i] != min_y else second_min_y
    cur_max_y = max_y if y[i] != max_y else second_max_y

    ans = min(ans, (cur_max_x - cur_min_x) * (cur_max_y - cur_min_y))

fout.write(f"{ans}\n")
fout.close()
